llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8702,61 +8738,264 @@ SDValue SystemZTargetLowering::combineSETCC(
+      CCVals.emplace_back(
+          (CCMaskValsSet.count(CCVal) && CCValidValsSet.count(CCVal))
+              ? TrueVals[CCVal]
+              : FalseVals[CCVal]);
+    return CCVals;
+  }
+  SDValue Op0, Op1;
+  unsigned Opcode;
+  if (isValidBinaryOperation(Val, Op0, Op1, Opcode)) {
+    auto &&CC0Vals = simplifyAssumingCCVal(Op0, CC);
+    auto &&CC1Vals = simplifyAssumingCCVal(Op1, CC);
+    SmallVector<int, 4> CCVals;
+    for (auto CCVal : {0, 1, 2, 3})
+      CCVals.emplace_back(BinaryOPS[Opcode](CC0Vals[CCVal], CC1Vals[CCVal]));
+    return CCVals;
+  }
+  return {};
+}
+
 static bool combineCCMask(SDValue &CCReg, int &CCValid, int &CCMask) {
   // We have a SELECT_CCMASK or BR_CCMASK comparing the condition code
   // set by the CCReg instruction using the CCValid / CCMask masks,
-  // If the CCReg instruction is itself a ICMP testing the condition
+  // If the CCReg instruction is itself a ICMP / TM  testing the condition
   // code set by some other instruction, see whether we can directly
   // use that condition code.
-
-  // Verify that we have an ICMP against some constant.
-  if (CCValid != SystemZ::CCMASK_ICMP)
-    return false;
-  auto *ICmp = CCReg.getNode();
-  if (ICmp->getOpcode() != SystemZISD::ICMP)
-    return false;
-  auto *CompareLHS = ICmp->getOperand(0).getNode();
-  auto *CompareRHS = dyn_cast<ConstantSDNode>(ICmp->getOperand(1));
-  if (!CompareRHS)
+  auto *CCNode = CCReg.getNode();
+  if (!CCNode)
     return false;
-
-  // Optimize the case where CompareLHS is a SELECT_CCMASK.
-  if (CompareLHS->getOpcode() == SystemZISD::SELECT_CCMASK) {
-    // Verify that we have an appropriate mask for a EQ or NE comparison.
-    bool Invert = false;
-    if (CCMask == SystemZ::CCMASK_CMP_NE)
-      Invert = !Invert;
-    else if (CCMask != SystemZ::CCMASK_CMP_EQ)
+  if (CCNode->getOpcode() == SystemZISD::TM) {
+    if (CCValid != SystemZ::CCMASK_TM)
       return false;
-
-    // Verify that the ICMP compares against one of select values.
-    auto *TrueVal = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(0));
-    if (!TrueVal)
+    if ((CCMask != SystemZ::CCMASK_TM_SOME_1) &&
+        (CCMask != SystemZ::CCMASK_TM_ALL_0))
       return false;
-    auto *FalseVal = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(1));
-    if (!FalseVal)
+    SDValue Op0 = CCNode->getOperand(0);
+    SDValue Op1 = CCNode->getOperand(1);
+    auto [Op0CC, Op0CCValid] = findCCUse(Op0);
+    if (Op0CC == SDValue())
       return false;
-    if (CompareRHS->getAPIntValue() == FalseVal->getAPIntValue())
-      Invert = !Invert;
-    else if (CompareRHS->getAPIntValue() != TrueVal->getAPIntValue())
+    int TMConstVal;
+    auto *N = dyn_cast<ConstantSDNode>(Op1);
+    if (N) {
+      TMConstVal = N->getZExtValue();
+      if ((TMConstVal != (1 << SystemZ::IPM_CC)) && (TMConstVal != 1))
+        return false;
+    }
+    auto &&Op0CCVals = simplifyAssumingCCVal(Op0, Op0CC);
+    auto [Op1CC, Op1CCValid] = findCCUse(Op1);
```

Comment:
This is not correct.  As above, we *must* decide on one single CC node for the `simplifyAssumingCCVal` optimization.  We are free how to choose this single node, but we can never mix two.  Here, I'd first look into Op0, and this is finds a CC, use it; otherwise, look into Op1 and use any CC found there.